WordPress/gutenberg
# Replace react-native-hr library

## What?
Replaces the forked version of [react-native-hr](https://github.com/wordpress-mobile/react-native-hr/) in favor of implementing the library within Gutenberg instead. 

**NOTE:** I started this PR without realizing that a similar effort to remove react-native-hr had already been undertaken in https://github.com/WordPress/gutenberg/pull/48986, which remains open. The approach in this PR is essentially a direct reimplementation of the original library code, which may address some of the feedback concerns from https://github.com/WordPress/gutenberg/pull/48986 regarding matching web more closely.

@dcalhoun and @geriux -- let's discuss and get one or the other merged either way, as we've now identified twice that this library is completely unnecessary as a third-party fork and can be removed. 😅 

## Why?
To reduce the number of third-party libraries and dependencies in order to support React Native's [New Architecture](https://reactnative.dev/docs/new-architecture-intro). 

## How?
The PR takes the source of our forked react-native-hr code, which is just [one single JS file](https://github.com/wordpress-mobile/react-native-hr/blob/master/src/index.js), and reimplements the source within the HorizontalRule primitive (while converting the code from a class component to a function component in the process). Then, the files that reference the primitive are updated with the new internal dependency.

## Testing Instructions
1. Create a new post
2. Add a Separator, Page Break, and More block
3. Switch from Light to Dark mode
4. Observe behavior from `trunk` (or production) is identical 

## Screenshots or screencast 
<img width="348" alt="Screenshot 2023-05-29 at 4 07 07 pm" src="https://github.com/WordPress/gutenberg/assets/643285/3f17b9e2-76a5-4f67-8de5-e92a7558eba9">
<img width="348" alt="Screenshot 2023-05-29 at 4 02 00 pm" src="https://github.com/WordPress/gutenberg/assets/643285/62c9ed89-0d06-48f0-8c65-248c22423e1e">
